T33110 fix shadow dom fouc

dylanmccall opened the pull request:
https://phabricator.endlessm.com/T33110

dylanmccall committed:
ZimRendererIndex: Fix container missing scrollbar

https://phabricator.endlessm.com/T33110

dylanmccall committed:
ZimContentView: Wait for resources to load before showing content

This change prevents a flash of unstyled content from appearing when a
new article is added to the DOM, before the related stylesheets are
loaded.

https://phabricator.endlessm.com/T33110

dylanmccall commented:
So this _almost_ works. The current implementation puts the new DOM tree a separate container, waits for it to load, then copies it to the main (visible) container. It only "almost" works, because there is still a flash when styles are newly reloaded (from cache) in their new context - but it is consistently brief, at least, because we know everything was already fetched a moment earlier. Instead, we need to split this into some more manageable components, so we can have a component that simply renders a DOM tree (_not_ a URL) and reports when it is finished, instead of mutating `ZimContentView`. With things split up more elegantly and connecting through props instead of internal state, we can hopefully change `ZimContentView` to use the [transitions](https://v2.vuejs.org/v2/guide/transitions.html) system we already have in Vue.

dylanmccall committed:
Add DOMTreeRenderer component

This contains the shadow DOM code previously in ZIMContentView, which
will make it easier to transition between pages.

https://phabricator.endlessm.com/T33110

dylanmccall committed:
Use a JSON API to load random articles

https://phabricator.endlessm.com/T33110

dylanmccall commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -0,0 +1,160 @@
+
+  @import '~kolibri-design-system/lib/styles/definitions';
+
+  .content {
+    display: block;
+    width: 100%;
+    height: 100%;
+    overflow: auto;
+    // Set the transform property so fixed-position children are relative to
+    // this element.
+    transform: translate(0);
+  }
```
This was all copied and pasted from `zimContentView`. It makes sense that this one is the scrolling container, but some other style rules should probably only be there (and vice versa).

danigm commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -0,0 +1,160 @@
+    components: {},
+    props: {
+      document: {
+        type: Element,
```
I'm getting some errors in the console related to this property type:

```
Invalid prop: type check failed for prop "document". Expected Element, got HTMLDocument 
```

danigm commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -0,0 +1,160 @@
+
+          this.shadow.replaceChildren(document.documentElement);
+
+          if (this.resourcesToLoad === 0 && this.resourcesLoaded === 0) {
```
nitpick: The second check in this `if` is not needed, I think it's enough to check the number of `resourcesToLoad`;

danigm commented on kolibri_zim_plugin/assets/src/views/ZimContentView.vue:
```diff
@@ -1,15 +1,15 @@
+    :document="document"
+    :location="location"
+    :openExternalLinksInNewWindow="true"
+    :class="[ 'content', documentReady === false ? 'content--loading' : null ]"
```
This can be done with a [different syntax](https://v2.vuejs.org/v2/guide/class-and-style.html#Object-Syntax):

```
class="content"
:class="{ 'content--loading': !documentReady }"
```

dylanmccall replied:
Ah, I was looking for that! Thank you :)

danigm commented on kolibri_zim_plugin/assets/src/views/ZimContentView.vue:
```diff
@@ -191,13 +170,15 @@
     height: 100%;
     overflow: auto;
+    opacity: 1;
+    transition: opacity 0.5s ease;
```
This transition should be removed, it shows the unstyled content to new content transition as a fade-out & fade-in, without the transtion works correctly for me in chromium, but with this line I can see the unstyled content before the correct presentation.

dylanmccall committed:
Fix incorrect prop type in DOMTreeRenderer

https://phabricator.endlessm.com/T33110

dylanmccall committed:
Remove unused styles in DOMTreeRenderer

https://phabricator.endlessm.com/T33110

dylanmccall committed:
Special case home pages so the correct breadcrumb appears

https://phabricator.endlessm.com/T33110

dylanmccall committed:
Show articles in unique DOMTreeRenderer instances

This change improves transitions between articles by switching focus
between multiple DOMTreeRenderer instances.

https://phabricator.endlessm.com/T33110

dylanmccall committed:
Show a spinner while articles are loading

https://phabricator.endlessm.com/T33110

manuq commented on kolibri_zim_plugin/assets/src/views/ZimRendererIndex.vue:
```diff
@@ -326,6 +326,7 @@
 
   .zim-content-view {
     z-index: 1;
+    height: 100%;
```
:+1:

manuq commented on kolibri_zim_plugin/assets/src/views/ZimContentView.vue:
```diff
@@ -87,11 +98,30 @@
+        this.resourcesLoaded += 1;
+        this.checkDocumentLoadFinished();
+      },
+      checkDocumentLoadFinished() {
```
Nit: This method is not only doing the check but also doing the replacement. So I would rename it.

manuq commented on kolibri_zim_plugin/assets/src/views/ZimContentView.vue:
```diff
@@ -191,13 +170,15 @@
 
-  .load-content {
-    display: none;
+  .content--loading {
```
Nit why 2 dashes in this class name? We don't have other cases in our source code.

manuq commented on kolibri_zim_plugin/views.py:
```diff
@@ -167,10 +167,8 @@ class ZimRandomArticleView(_ZimFileViewMixin, View):
-            request, zim_filename, self.zim_file.random_article_url
-        )
-        return HttpResponseRedirect(article_url)
+        random_article_path = self.zim_file.random_article_url
+        return JsonResponse({"zimFile": zim_filename, "zimPath": random_article_path})
```
How is this JSON API related to the flashing of unstyled content? It seems like a fix more related to the other ticket https://github.com/endlessm/kolibri-zim-plugin/pull/15

manuq commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -12,7 +12,7 @@
     props: {
       document: {
-        type: Element,
+        type: Document,
```
:+1:

manuq commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -141,20 +141,10 @@
-    display: block;
-    width: 100%;
-    height: 100%;
-    overflow: auto;
```
:+1:

manuq commented on kolibri_zim_plugin/assets/src/views/DOMTreeRenderer.vue:
```diff
@@ -141,20 +141,10 @@
 
-  .load-content {
-    display: none;
-  }
```
This is undoing a change done in this very PR. So it deserves squashing with the previous commit.

manuq commented on kolibri_zim_plugin/views.py:
```diff
@@ -137,6 +139,16 @@ def get(self, request, zim_filename, zim_article_path):
             )
             return HttpResponseRedirect(article_url)
 
+        if self._article_is_main_page(zim_article) and redirect_from != "":
+            # Make the main page article work like it does via ZimIndexView
+            article_url = _zim_article_url(
+                request,
+                zim_filename,
+                zim_article.full_url,
+                redirect_from="",
+            )
+            return HttpResponseRedirect(article_url)
```
:+1:

manuq commented on kolibri_zim_plugin/assets/src/views/ZimContentView.vue:
```diff
@@ -103,6 +103,7 @@
         return this.pushArticleFromUrl(articleId, requestUrl, zimHash);
       },
       pushArticleFromUrl(articleId, requestUrl, zimHash) {
+        this.$emit('articleLoading');
```
:+1:
